---
title: "Script_for_WAT_Data_Preparation"
author: "Yuxin Wang"
output: 
  html_document: 
    fig_height: 10
    fig_width: 10
editor_options: 
  chunk_output_type: console
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(fig.keep = "high")
```

# library

```{r library}
library(rtracklayer)
library(data.table)
library(limma)
library(FactoMineR)
library(factoextra)
library(sva)
library(GEOquery)
library(org.Hs.eg.db)
library(tidyverse)
```

# GTEx adipose RNA-seq samples

# we downloaded data from the GTEx portal (https://www.gtexportal.org). data information are as follows:
# GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt
# GTEx_Analysis_v8_Annotations_SubjectPhenotypesDS.txt
# GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_reads.gct

# we downloaded gene annotation file from GENCODE (https://www.gencodegenes.org/)
# gencode.v26.annotation.gtf.gz

```{r GTEx}
# global options
rm(list = ls())
options(stringsAsFactors = FALSE)

# load annotation data
gtf <- rtracklayer::import("./rawData/GTEx/gencode.v26.annotation.gtf.gz") %>%
  as.data.frame() %>%
  subset(select = c("gene_id", "gene_name", "gene_type")) %>%
  subset(subset = grepl(.$gene_type, pattern = "protein_coding")) %>%
  subset(subset = !duplicated(.))

# isolate protein codingn genes
PC <- unique(gtf$gene_id)

# load attribution data
attribute <- read.table(file = "./rawData/GTEx/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt", 
                        sep = '\t', 
                        quote = "", 
                        fill = T,
                        header = T) %>%
  subset(subset = .$SMTS == "Adipose Tissue") %>%
  subset(subset = .$SMGEBTCHT == "TruSeq.v1") %>%
  subset(subset = .$SMAFRZE == "RNASEQ")

# isolate useful information
attribute <- data.frame(ID = str_split(attribute$SAMPID, "-", simplify = T)[, 2],
                        sample = attribute$SAMPID, 
                        tissue = attribute$SMTS, 
                        group = attribute$SMTSD,
                        SMRIN = attribute$SMRIN, 
                        SMNABTCHT = attribute$SMNABTCHT,
                        SMCENTER = attribute$SMCENTER)

# load donor information
phenoType <- read.table(file = "./rawData/GTEx/GTEx_Analysis_v8_Annotations_SubjectPhenotypesDS.txt", 
                        sep = '\t', 
                        quote = "", 
                        fill = T,
                        header = T,
                        check.names = F)
phenoType$ID <- str_split(phenoType$SUBJID, "-", simplify = T)[, 2]

# merge attribute and phenoType
phenoData <- merge(attribute, phenoType, by = "ID")

# isolate information of adipose tissue
phenoData <- phenoData[phenoData$tissue == "Adipose Tissue",]

# rename group information
phenoData$group <- gsub("Adipose - Subcutaneous", "SAT", phenoData$group)
phenoData$group <- gsub("Adipose - Visceral.*", "VAT", phenoData$group)
phenoData$SEX <- gsub("1", "Male", phenoData$SEX)
phenoData$SEX <- gsub("2", "Female", phenoData$SEX)

# rename unknown phenotype information
phenoData$SMCENTER[phenoData$SMCENTER == ""] = "Unknown"

# rename unknown phenotype information
phenoData$DTHHRDY[is.na(phenoData$DTHHRDY)] = 5

# group RNA integrity according to quantile
quantile_RIN <- quantile(phenoData$SMRIN)
phenoData$quantile_RIN <- ifelse(phenoData$SMRIN >= quantile_RIN[3],
              ifelse(phenoData$SMRIN >= quantile_RIN[4], 4, 3),
              ifelse(phenoData$SMRIN >= quantile_RIN[2], 2, 1))

phenoData <- arrange(phenoData, group, ID)

# load read counts
counts <- fread(file = "./rawData/GTEx/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_reads.gct", 
                   sep = "\t", 
                   skip = 2, 
                   select = c("Name", phenoData$sample)) %>%
  subset(subset = .$Name %in% PC) %>% 
  as.data.frame() %>% 
  column_to_rownames("Name") %>% 
  as.matrix()

dim(counts)

# arrange phenoData
phenoData <- phenoData[match(colnames(counts), phenoData$sample),]
phenoData$sample[1:10]
colnames(counts)[1:10]

# arrange gtf
gtf <- gtf[match(rownames(counts), gtf$gene_id), ]
gtf$gene_id[1:10]
rownames(counts)[1:10]

# remove duplicated genes
dim(counts)
keep <- by(counts, gtf$gene_name, function(x) rownames(x)[which.max(rowMeans(x))]) %>% as.character()
counts <- counts[keep, ]
dim(counts)

# rearrange gtf
gtf <- gtf[match(rownames(counts), gtf$gene_id), ]
gtf[1:10, ]
counts[1:10,1:2]

# remove batch effect
# calculate counts per million
cps <- apply(counts, 2, function(x) 1e+06 * (x/sum(x)))

# log transformation
lcps <- log2(cps+1)

# plot PCA for visualizing systematic influences of particular covariates on the dispersion of adipose expression data.
dat <- as.data.frame(t(lcps))
pca <- PCA(dat, graph = FALSE)

# function for PCA plots
plotMyPCA <- function(pca, title, group){
  require(FactoMineR)
  require(factoextra)
  p <- fviz_pca_ind(pca, 
                    label = "none", 
                    title =  title,
                    pointsize = 1,
                    pointshape = 19,
                    addEllipses = T, 
                    ellipse.level = 0.95, 
                    mean.point = FALSE,
                    habillage = group)
  p+theme(plot.title = element_text(hjust = 0.5))
}

# analyze the effects of tissue type (biological influence)
plotMyPCA(pca, "Tissue type", as.factor(phenoData$group))

# analyze the effects of sex (biological influence)
plotMyPCA(pca, "Sex", as.factor(phenoData$SEX))

# analyze the effects of age (biological influence)
plotMyPCA(pca, "Age", as.factor(phenoData$AGE))

# analyze the effects of death circumstances (biological influence)
plotMyPCA(pca, "Hardy scale", as.factor(phenoData$DTHHRDY))

# analyze the effects of RNA integrity (batch influence)
plotMyPCA(pca, "RNA integrity", as.factor(phenoData$quantile_RIN))

# analyze the effects of collection site (batch influence)
plotMyPCA(pca, "Collection site", as.factor(phenoData$SMCENTER))

# analyze the effects of extraction method (batch influence)
plotMyPCA(pca, "Extraction method", as.factor(phenoData$SMNABTCHT))

# notice that no batch effect significantly influences the dispersion of adipose expression data
# thus, no batch effect is corrected in these data

# save data for further analysis
save(counts, gtf, phenoData, file = "./RData/cohort_GTEx_AT.RData")

```

# we downloaded data from Gene Expression Omnibus (GEO) repository: 
# GSE135134_METSIM_subcutaneousAdipose_RNAseq_TPMs_n434.txt.gz

```{r}
# global options
rm(list = ls())
options(stringsAsFactors = FALSE)

# load data
tpms <- read.table(file = "./rawData/GSE135134/GSE135134_METSIM_subcutaneousAdipose_RNAseq_TPMs_n434.txt.gz", sep = "\t", header = T)

# load annotation data
gtf <- rtracklayer::import("./rawData/GSE135134/gencode.v19.annotation.gtf.gz") %>%
  as.data.frame() %>%
  subset(select = c("gene_id", "gene_name", "gene_type")) %>%
  subset(subset = grepl(.$gene_type, pattern = "protein_coding")) %>%
  subset(subset = !duplicated(.))

# remove duplicated genes
tpms <- tpms[tpms$ID %in% gtf$gene_id,] %>% remove_rownames() %>% column_to_rownames("ID")
gtf <- gtf[match(rownames(tpms), gtf$gene_id), ]
dim(tpms)
tpms <- avereps(tpms, ID = gtf$gene_name)
dim(tpms)

# rearrange gtf
gtf <- gtf[match(rownames(tpms), gtf$gene_name), ]

# download smaple annotation informations
getPhenoData <- function(GSE){
  require(GEOquery)
  getGEOfile(GSE, destdir = ".")
  eset <- getGEO(filename = paste0(GSE, ".soft.gz"))
  eset <- eset@gsms
  
  file.remove(paste0(GSE, ".soft.gz"))
  
  phenoData <- data.frame()
  
  for (i in 1:length(eset)) {
    gsm <- eset[[i]] %>% .@header 
    gsm <- c(gsm$title, gsm$characteristics_ch1)
    phenoData <- rbind(phenoData, gsm)
  }
  colnames(phenoData) <- c("sample", "gender", "batch", "integrity", "age", "bmi")
  
  phenoData
}

phenoData <- getPhenoData(GSE = "GSE135134")

phenoData$sample <- paste0("X", phenoData$sample)
phenoData$gender <- gsub(".*: ", "", phenoData$gender)
phenoData$batch <- gsub(".*: ", "", phenoData$batch)
phenoData$batch <- gsub(" ", "_", phenoData$batch)
phenoData$batch <- factor(phenoData$batch)
phenoData$integrity <- gsub(".*: ", "", phenoData$integrity)
phenoData$age <- gsub(".*: ", "", phenoData$age)
phenoData$bmi <- gsub(".*: ", "", phenoData$bmi)
phenoData$group <- ifelse(phenoData$bmi < 18.5, "Underweight", 
                          ifelse(phenoData$bmi < 25, "Normal", 
                                 ifelse(phenoData$bmi < 30, "Overweight", "Obese")))

phenoData <- phenoData[match(phenoData$sample, colnames(tpms)),]

# visualize batch effects
ltpms <- log2(tpms+1)
dat <- as.data.frame(t(ltpms))
pca <- PCA(dat, graph = FALSE)

# function for generating annotated PCA plots
plotMyPCA <- function(pca, title, group){
  require(FactoMineR)
  require(factoextra)
  p <- fviz_pca_ind(pca, 
                    label = "none", 
                    title =  title,
                    pointsize = 1,
                    pointshape = 19,
                    addEllipses = T, 
                    ellipse.level = 0.95, 
                    mean.point = FALSE,
                    habillage = group)
  p+theme(plot.title = element_text(hjust = 0.5))
}

# analyze the effects of batch
plotMyPCA(pca, "sequencing batch", phenoData$batch)

# remove unqualified samples
# delete batch_1 and batch_4 samples
phenoData <- phenoData[phenoData$batch %in% c("batch_2", "batch_3"),]

ltpms <- ltpms[,phenoData$sample]

save(ltpms, gtf, phenoData, file = "./RData/cohort_GSE135134.RData")

```

